photon-ml/src/main/scala/com/linkedin/photon/ml/data/DataReader.scala

```diff
@@ -0,0 +1,158 @@
+   *
+   * @param paths the paths to the files or folders
+   * @param featureColumnMap a map that specifies how the feature columns should be merged. The keys specify the name
+   *   of the merged destination column, and the values are sets of source columns to merge, e.g.:
+   *
+   *     Map("userFeatures" -> Set("profileFeatures", "titleFeatures"))
+   *
+   *   This configuration merges the "profileFeatures" and "titleFeatures" columns into a single column named
+   *   "userFeatures".
+   * @return the loaded and transformed DataFrame
+   */
+  def read(
+      paths: Seq[String],
+      featureColumnMap: Map[String, Set[String]]): (DataFrame, Map[String, IndexMapLoader]) = {
+
+    val (first, indexMapLoaders) = read(paths.head, featureColumnMap)
+    val df = paths.tail.foldLeft(first) { case (acc: DataFrame, path: String) =>
+      val (curr, _) = read(path, featureColumnMap)
+      acc.unionAll(curr)
+    }
+
+    (df, indexMapLoaders)
+  }
+
+  /**
+   * Reads the files at the given paths into a DataFrame, using the given index map for feature names
+   *
+   * @param paths the paths to the files or folders
+   * @param indexMapLoaders a map of index map loaders, containing one loader for each feature column
+   * @param featureColumnMap a map that specifies how the feature columns should be merged. The keys specify the name
+   *   of the merged destination column, and the values are sets of source columns to merge, e.g.:
+   *
+   *     Map("userFeatures" -> Set("profileFeatures", "titleFeatures"))
+   *
+   *   This configuration merges the "profileFeatures" and "titleFeatures" columns into a single column named
+   *   "userFeatures".
+   * @return the loaded and transformed DataFrame
+   */
+  def read(
+      paths: Seq[String],
+      indexMapLoaders: Map[String, IndexMapLoader],
+      featureColumnMap: Map[String, Set[String]]): DataFrame = {
+
+    val first = read(paths.head, indexMapLoaders, featureColumnMap)
+    val df = paths.tail.foldLeft(first) { case (acc: DataFrame, path: String) =>
+      acc.unionAll(read(path, indexMapLoaders, featureColumnMap))
+    }
+
+    df
+  }
+
+  /**
+   * Reads the files at the given paths into a DataFrame, assuming the default feature vector
+   *
+   * @param paths the paths to the files or folders
+   * @return the loaded and transformed DataFrame
+   */
+  def read(paths: Seq[String]): (DataFrame, IndexMapLoader) = {
+    val (data, indexMapLoaders) = read(paths, defaultFeatureColumnMap)
+    (data, indexMapLoaders(defaultFeatureColumn))
+  }
+
+  /**
+   * Reads the files at the given paths into a DataFrame, using the given index map for feature names
+   *
+   * @param paths the paths to the files or folders
+   * @param indexMapLoaders a map of index map loaders, containing one loader for each feature column
+   * @return the loaded and transformed DataFrame
+   */
+  def read(paths: Seq[String], indexMapLoaders: Map[String, IndexMapLoader]): DataFrame =
```

Comment:
Ah, you're right. It's using the return type to disambiguate the call, but that is a little awkward. I think I'll rename the functions that merge columns to `readMerged`. This makes the combined transformation a little more explicit anyway. Sound reasonable?